src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -1058,6 +1092,204 @@ internal static void PerformWaitOrTimerCallback(_ThreadPoolWaitOrTimerCallback h
         }
     }
 
+    internal class ThreadPoolBlockingQueue : IThreadPoolWorkItem
+    {
+        [ThreadStatic]
+        private static object? t_currentWorkItem;
+
+        private static ThreadPoolBlockingQueue? s_processor;
+        private static HashSet<object>? s_blockingItems;
```

Comment:
Is there any concern about how big this can get? I guess it's a function of the code since we're looking at types not work item instances.

Reply:
Its only recording the "methods" (which includes statemachine `Type`s for async methods) that initiate the blocking call stack so hopefully should be of small size; if not then the app will have bigger issues...

Reply:
Are OS scheduler hiccups or little bursts of activity end up adding innocent victims to this list? Can this lead to close to everything that the app runs being in this list eventually, once the app runs for a while?

Reply:
The code will need to be at some point explicitly making a call to `Task.Wait()`, `Task.Result`, `Task.WaitAll(...)`, `.GetAwaiter().GetResult()` (or `ValueTask` variants) that doesn't have a result; so not sure they are innocent 😉

Regular OS blocking/sync calls that don't block a Task shouldn't effect it (e.g. `.Read()`); but that's not so problematic as they will eventually unblock, whereas "sync-over-async"; which this aims to address, needs a second thread to unblock and if there's no spare on the ThreadPool and they are being blocked as fast as they are being injected they will never unblock.

The main down side is it marks the entry point from the last yield (as it doesn't break the stack); so anything branching off that same yield point will be effected; until it itself yields when it will go back to a regular ThreadPool item.

However it does mean mitigation patterns like:
```csharp
await Task.Run(() => BlockingCall());
```
Will actually do something; scoping the ThreadPool mitigation at the lambda level rather than further back; and adding a yield
```csharp
await Task.Yield();
BlockingCall();
```
Will again scope to the mitigation to async method rather than further back in the stack.

This could be for example calling Azure KeyVault async methods via the `System.Security.Cryptography` classes where they have no async apis; so would need to be `.GetAwaiter.GetResult()` behind the scenes.

Reply:
I think we should think a bit about what we can do to make the finger printing a little better (can we do anything at JIT time or runtime) and try to figure out if we need to have a heuristic with how to remove these blocking call sites from this list. That can be improved over time (I can think of some crazy ideas...) but we just need to make sure that the idea is sound (I think it is).

It's already currently scoped to sync over async which is think is the big thing we need to tackle.

Reply:
I wonder if we can self heal these blocking call sites by keeping some metrics of blocking call site execution and removing it from the list if multiple executions complete within some threshold (spitballing here) or don't make the same blocking call on re-execution.